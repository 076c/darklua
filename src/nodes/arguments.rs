use crate::lua_generator::{LuaGenerator, ToLua};
use crate::nodes::{
    Expression,
    StringExpression,
    TableExpression,
};

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum Arguments {
    Tuple(Vec<Expression>),
    String(StringExpression),
    Table(TableExpression),
}

impl Arguments {
    pub fn to_expressions(self) -> Vec<Expression> {
        match self {
            Self::Tuple(expressions) => expressions,
            Self::String(string) => vec![string.into()],
            Self::Table(table) => vec![table.into()],
        }
    }

    pub fn append_argument<T: Into<Expression>>(self, argument: T) -> Self {
        let mut expressions = self.to_expressions();
        expressions.push(argument.into());
        Arguments::Tuple(expressions)
    }
}

impl Default for Arguments {
    fn default() -> Self {
        Arguments::Tuple(Vec::new())
    }
}

impl ToLua for Arguments {
    fn to_lua(&self, generator: &mut LuaGenerator) {
        match self {
            Self::String(string) => string.to_lua(generator),
            Self::Table(table) => table.to_lua(generator),
            Self::Tuple(expressions) => {
                generator.merge_char('(');

                let last_index = expressions.len().checked_sub(1).unwrap_or(0);
                expressions.iter().enumerate()
                    .for_each(|(index, expression)| {
                        expression.to_lua(generator);

                        if index != last_index {
                            generator.push_char(',');
                        }
                    });

                generator.push_char(')');
            }
        }
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn generate_tuple_does_not_break_on_new_line() {
        let mut generator = LuaGenerator::new(16);
        generator.push_str("123");
        generator.push_str("functionName");

        Arguments::Tuple(Vec::new()).to_lua(&mut generator);

        assert_eq!(generator.into_string(), "123\nfunctionName()");
    }

    mod snapshot {
        use super::*;

        use insta::assert_snapshot;

        fn get_expression() -> Expression {
            Expression::Identifier("foo".to_owned()).into()
        }

        #[test]
        fn empty_tuple() {
            assert_snapshot!("empty_tuple", Arguments::Tuple(vec![]).to_lua_string());
        }

        #[test]
        fn single_argument() {
            let arguments = Arguments::Tuple(vec![get_expression()]);

            assert_snapshot!("single_argument", arguments.to_lua_string());
        }

        #[test]
        fn two_arguments() {
            let arguments = Arguments::Tuple(vec![
                get_expression(),
                get_expression(),
            ]);

            assert_snapshot!("two_arguments", arguments.to_lua_string());
        }
    }
}
