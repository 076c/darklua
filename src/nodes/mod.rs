//! The collection of nodes used for the Lua abstract syntax tree.

mod arguments;
mod block;
mod expressions;
mod function_call;
mod statements;

pub use arguments::*;
pub use block::*;
pub use expressions::*;
pub use function_call::*;
pub use statements::*;
