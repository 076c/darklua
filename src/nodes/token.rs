#[derive(Clone, Debug, PartialEq, Eq)]
pub enum Position {
    Reference { start: usize, end: usize },
    LineNumber { content: String, line_number: usize },
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum TriviaKind {
    Comment,
    Whitespace,
}

impl TriviaKind {
    pub fn at(self, start: usize, end: usize) -> Trivia {
        Trivia {
            position: Position::Reference { start, end },
            kind: self,
        }
    }
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Trivia {
    position: Position,
    kind: TriviaKind,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Token {
    position: Position,
    leading_trivia: Vec<Trivia>,
    trailing_trivia: Vec<Trivia>,
}

impl Token {
    pub fn new(start: usize, end: usize) -> Self {
        Self {
            position: Position::Reference { start, end },
            leading_trivia: Vec::new(),
            trailing_trivia: Vec::new(),
        }
    }

    pub fn with_leading_trivia(mut self, trivia: Trivia) -> Self {
        self.leading_trivia.push(trivia);
        self
    }

    pub fn with_trailing_trivia(mut self, trivia: Trivia) -> Self {
        self.trailing_trivia.push(trivia);
        self
    }

    #[inline]
    pub fn push_leading_trivia(&mut self, trivia: Trivia) {
        self.leading_trivia.push(trivia);
    }

    #[inline]
    pub fn push_trailing_trivia(&mut self, trivia: Trivia) {
        self.trailing_trivia.push(trivia);
    }
}

#[cfg(test)]
mod test {
    // use super::*;

    #[test]
    fn do_test() {
        // struct A {
        //     start: usize,
        //     end: usize,
        // }
        // struct B {
        //     content: String,
        //     // line_number: Option<usize>,
        //     line_number: usize,
        // }
        // assert_eq!(4, std::mem::size_of::<Position>());
        // assert_eq!(std::mem::size_of::<A>(), std::mem::size_of::<B>());
    }
}
