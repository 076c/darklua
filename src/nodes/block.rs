use crate::lua_generator::{LuaGenerator, ToLua};
use crate::nodes::{LastStatement, Statement};

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Block {
    statements: Vec<Statement>,
    last_statement: Option<LastStatement>,
}

impl Block {
    pub fn new(statements: Vec<Statement>, last_statement: Option<LastStatement>) -> Self {
        Self {
            statements,
            last_statement,
        }
    }

    pub fn with_statement<T: Into<Statement>>(mut self, statement: T) -> Self {
        self.statements.push(statement.into());
        self
    }

    pub fn with_last_statement(mut self, last_statement: LastStatement) -> Self {
        self.last_statement = Some(last_statement);
        self
    }

    pub fn is_empty(&self) -> bool {
        self.last_statement.is_none() && self.statements.is_empty()
    }

    pub fn get_statements(&self) -> &Vec<Statement> {
        &self.statements
    }

    pub fn filter_statements<F>(&mut self, mut f: F) where F: FnMut(&mut Statement) -> bool {
        let mut i = 0;

        while i != self.statements.len() {
            if f(&mut self.statements[i]) {
                i += 1;
            } else {
                self.statements.remove(i);
            }
        }
    }

    pub fn mutate_statements(&mut self) -> &mut Vec<Statement> {
        &mut self.statements
    }

    pub fn mutate_last_statement(&mut self) -> &mut Option<LastStatement> {
        &mut self.last_statement
    }
}

impl Default for Block {
    fn default() -> Self {
        Self::new(Vec::new(), None)
    }
}

impl ToLua for Block {
    fn to_lua(&self, generator: &mut LuaGenerator) {
        generator.for_each_and_between(
            &self.statements,
            |generator, statement| statement.to_lua(generator),
            |generator| generator.push_char(';'),
        );

        if let Some(last_statement) = &self.last_statement {
            if self.statements.len() > 0 {
                generator.push_char(';');
            };

            last_statement.to_lua(generator);
            generator.push_char(';');
        }
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::nodes::DoStatement;

    #[test]
    fn default_block_is_empty() {
        let block = Block::default();

        assert!(block.is_empty());
    }

    #[test]
    fn is_empty_is_true_when_block_has_no_statements_or_last_statement() {
        let block = Block::new(Vec::new(), None);

        assert!(block.is_empty());
    }

    #[test]
    fn is_empty_is_false_when_block_has_a_last_statement() {
        let block = Block::default().with_last_statement(LastStatement::Break);

        assert!(!block.is_empty());
    }

    #[test]
    fn is_empty_is_false_when_block_a_statement() {
        let block = Block::default().with_statement(DoStatement::default());

        assert!(!block.is_empty());
    }
}
