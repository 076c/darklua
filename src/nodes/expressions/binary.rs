use crate::lua_generator::{LuaGenerator, ToLua};
use crate::nodes::Expression;

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum BinaryOperator {
    And,
    Or,
    Equal,
    NotEqual,
    LowerThan,
    LowerOrEqualThan,
    GreatherThan,
    GreatherOrEqualThan,
    Plus,
    Minus,
    Asterisk,
    Slash,
    Percent,
    Caret,
    Concat,
}

impl ToLua for BinaryOperator {
    fn to_lua(&self, generator: &mut LuaGenerator) {
        match self {
            Self::And => generator.push_str("and"),
            Self::Or => generator.push_str("or"),
            Self::Equal => generator.push_str("=="),
            Self::NotEqual => generator.push_str("~="),
            Self::LowerThan => generator.push_char('<'),
            Self::LowerOrEqualThan => generator.push_str("<="),
            Self::GreatherThan => generator.push_char('>'),
            Self::GreatherOrEqualThan => generator.push_str(">="),
            Self::Plus => generator.push_char('+'),
            Self::Minus => generator.push_char('-'),
            Self::Asterisk => generator.push_char('*'),
            Self::Slash => generator.push_char('/'),
            Self::Percent => generator.push_char('%'),
            Self::Caret => generator.push_char('^'),
            Self::Concat => generator.push_str(".."),
        }
    }
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct BinaryExpression {
    operator: BinaryOperator,
    left: Expression,
    right: Expression,
}

impl BinaryExpression {
    pub fn new(operator: BinaryOperator, left: Expression, right: Expression) -> Self {
        Self {
            operator,
            left,
            right,
        }
    }

    pub fn mutate_left(&mut self) -> &mut Expression {
        &mut self.left
    }

    pub fn mutate_right(&mut self) -> &mut Expression {
        &mut self.right
    }

    pub fn left(&self) -> &Expression {
        &self.left
    }

    pub fn right(&self) -> &Expression {
        &self.right
    }

    pub fn operator(&self) -> BinaryOperator {
        self.operator
    }
}

impl ToLua for BinaryExpression {
    fn to_lua(&self, generator: &mut LuaGenerator) {
        self.left.to_lua(generator);
        self.operator.to_lua(generator);
        self.right.to_lua(generator);
    }
}

#[cfg(test)]
mod test {
    use super::*;

    mod snapshot {
        use super::*;

        use insta::assert_snapshot;

        #[test]
        fn and_expression() {
            let expression = BinaryExpression::new(
                BinaryOperator::And,
                Expression::True,
                Expression::False
            );

            assert_snapshot!("and_expression", expression.to_lua_string());
        }

        #[test]
        fn equal_expression() {
            let expression = BinaryExpression::new(
                BinaryOperator::Equal,
                Expression::True,
                Expression::False
            );

            assert_snapshot!("equal_expression", expression.to_lua_string());
        }
    }
}
