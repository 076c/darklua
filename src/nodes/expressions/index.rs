use crate::nodes::{
    Expression,
    Prefix,
};
use crate::lua_generator::{LuaGenerator, ToLua};

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct IndexExpression {
    prefix: Prefix,
    index: Expression,
}

impl IndexExpression {
    pub fn new<E: Into<Expression>>(prefix: Prefix, expression: E) -> Self {
        Self {
            prefix,
            index: expression.into(),
        }
    }

    pub fn mutate_prefix(&mut self) -> &mut Prefix {
        &mut self.prefix
    }

    pub fn mutate_index(&mut self) -> &mut Expression {
        &mut self.index
    }
}

impl ToLua for IndexExpression {
    fn to_lua(&self, generator: &mut LuaGenerator) {
        self.prefix.to_lua(generator);
        generator.push_char('[');
        self.index.to_lua(generator);
        generator.push_char(']');
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn generate_identifier() {
        let output = IndexExpression::new(
            Prefix::from_name("foo"),
            Prefix::from_name("bar"),
        ).to_lua_string();

        assert_eq!(output, "foo[bar]");
    }
}
