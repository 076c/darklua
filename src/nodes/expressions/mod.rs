mod binary;
mod field;
mod function;
mod index;
mod number;
mod prefix;
mod string;
mod table;
mod unary;

pub use binary::*;
pub use field::*;
pub use function::*;
pub use index::*;
pub use number::*;
pub use prefix::*;
pub use string::*;
pub use table::*;
pub use unary::*;

use crate::lua_generator::{LuaGenerator, ToLua};
use crate::nodes::FunctionCall;

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum Expression {
    Binary(Box<BinaryExpression>),
    Call(Box<FunctionCall>),
    False,
    Field(Box<FieldExpression>),
    Function(FunctionExpression),
    Identifier(String),
    Index(Box<IndexExpression>),
    Nil,
    Number(NumberExpression),
    Parenthese(Box<Expression>),
    String(StringExpression),
    Table(TableExpression),
    True,
    Unary(Box<UnaryExpression>),
    VariableArguments,
}

impl From<BinaryExpression> for Expression {
    fn from(binary: BinaryExpression) -> Expression {
        Expression::Binary(Box::new(binary))
    }
}

impl From<FunctionCall> for Expression {
    fn from(call: FunctionCall) -> Expression {
        Expression::Call(Box::new(call))
    }
}

impl From<FieldExpression> for Expression {
    fn from(field: FieldExpression) -> Expression {
        Expression::Field(Box::new(field))
    }
}

impl From<FunctionExpression> for Expression {
    fn from(function: FunctionExpression) -> Self {
        Expression::Function(function)
    }
}

impl From<IndexExpression> for Expression {
    fn from(index: IndexExpression) -> Self {
        Self::Index(Box::new(index))
    }
}

impl From<NumberExpression> for Expression {
    fn from(number: NumberExpression) -> Self {
        Self::Number(number)
    }
}

impl From<Prefix> for Expression {
    fn from(prefix: Prefix) -> Self {
        match prefix {
            Prefix::Call(call) => Self::Call(Box::new(call)),
            Prefix::Field(field) => Self::Field(field),
            Prefix::Identifier(name) => Self::Identifier(name),
            Prefix::Index(index) => Self::Index(index),
            Prefix::Parenthese(expression) => Self::Parenthese(Box::new(expression)),
        }
    }
}

impl From<StringExpression> for Expression {
    fn from(string: StringExpression) -> Self {
        Self::String(string)
    }
}

impl From<TableExpression> for Expression {
    fn from(table: TableExpression) -> Self {
        Self::Table(table)
    }
}

impl From<UnaryExpression> for Expression {
    fn from(unary: UnaryExpression) -> Self {
        Self::Unary(Box::new(unary))
    }
}

fn break_variable_arguments(last_string: &str) -> bool {
    if let Some('.') = last_string.chars().last() {
        true
    } else if let Some(first_char) = last_string.chars().next() {
        first_char == '.' || first_char.is_digit(10)
    } else {
        false
    }
}

impl ToLua for Expression {
    fn to_lua(&self, generator: &mut LuaGenerator) {
        match self {
            Self::Binary(binary_expression) => binary_expression.to_lua(generator),
            Self::Call(call) => call.to_lua(generator),
            Self::False => generator.push_str("false"),
            Self::Field(field) => field.to_lua(generator),
            Self::Function(function) => function.to_lua(generator),
            Self::Identifier(identifier) => generator.push_str(identifier),
            Self::Index(index) => index.to_lua(generator),
            Self::Nil => generator.push_str("nil"),
            Self::Number(number) => number.to_lua(generator),
            Self::Parenthese(expression) => {
                generator.push_char('(');
                expression.to_lua(generator);
                generator.push_char(')');
            }
            Self::String(string) => string.to_lua(generator),
            Self::Table(table) => table.to_lua(generator),
            Self::True => generator.push_str("true"),
            Self::Unary(unary_expression) => unary_expression.to_lua(generator),
            Self::VariableArguments => {
                generator.push_str_and_break_if("...", break_variable_arguments);
            }
        }
    }
}

#[cfg(test)]
mod test {
    use super::*;

    mod to_lua {
        use super::*;

        #[test]
        fn generate_false_expression() {
            let output = Expression::False.to_lua_string();

            assert_eq!(output, "false");
        }

        #[test]
        fn generate_nil_expression() {
            let output = Expression::Nil.to_lua_string();

            assert_eq!(output, "nil");
        }

        #[test]
        fn generate_parenthese_expression() {
            let inner_expression = Box::new(Expression::True);
            let output = Expression::Parenthese(inner_expression).to_lua_string();

            assert_eq!(output, "(true)");
        }

        #[test]
        fn generate_true_expression() {
            let output = Expression::True.to_lua_string();

            assert_eq!(output, "true");
        }

        #[test]
        fn generate_variable_arguments_expression() {
            let output = Expression::VariableArguments.to_lua_string();

            assert_eq!(output, "...");
        }
    }
}
