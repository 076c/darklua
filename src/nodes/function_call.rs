use crate::lua_generator::{LuaGenerator, ToLua};
use crate::nodes::{
    Arguments,
    Prefix,
};

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct FunctionCall {
    prefix: Box<Prefix>,
    arguments: Arguments,
    method: Option<String>,
}

impl FunctionCall {
    pub fn new(prefix: Prefix, arguments: Arguments, method: Option<String>) -> Self {
        Self {
            prefix: Box::new(prefix),
            arguments,
            method,
        }
    }

    pub fn from_name<T: Into<String>>(name: T) -> Self {
        Self {
            prefix: Box::new(Prefix::Identifier(name.into())),
            arguments: Arguments::Tuple(Vec::new()),
            method: None,
        }
    }

    pub fn with_arguments(mut self, arguments: Arguments) -> Self {
        self.arguments = arguments;
        self
    }

    pub fn with_method(mut self, method: String) -> Self {
        self.method.replace(method);
        self
    }

    pub fn get_arguments(&self) -> &Arguments {
        &self.arguments
    }

    pub fn get_method(&self) -> Option<&String> {
        self.method.as_ref()
    }

    pub fn get_prefix(&self) -> &Prefix {
        &self.prefix
    }

    pub fn take_method(&mut self) -> Option<String> {
        self.method.take()
    }

    pub fn set_arguments(&mut self, arguments: Arguments) {
        self.arguments = arguments;
    }

    pub fn set_method(&mut self, method: String) {
        self.method.replace(method);
    }

    pub fn mutate_arguments(&mut self) -> &mut Arguments {
        &mut self.arguments
    }

    pub fn mutate_prefix(&mut self) -> &mut Prefix {
        &mut self.prefix
    }
}

impl ToLua for FunctionCall {
    fn to_lua(&self, generator: &mut LuaGenerator) {
        self.prefix.to_lua(generator);

        if let Some(method) = &self.method {
            generator.push_char(':');
            generator.push_str(&method);
        }

        self.arguments.to_lua(generator);
    }
}
