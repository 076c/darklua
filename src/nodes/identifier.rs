use crate::nodes::Token;

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Identifier {
    name: String,
    token: Option<Token>,
}

impl Identifier {
    pub fn new<S: Into<String>>(name: S) -> Self {
        Self {
            name: name.into(),
            token: None,
        }
    }

    pub fn with_token(mut self, token: Token) -> Self {
        self.token = Some(token);
        self
    }

    #[inline]
    pub fn set_token(&mut self, token: Token) {
        self.token = Some(token);
    }

    #[inline]
    pub fn get_name(&self) -> &String {
        &self.name
    }

    #[inline]
    pub fn mutate_name(&mut self) -> &mut String {
        &mut self.name
    }

    #[inline]
    pub fn set_name<IntoString: Into<String>>(&mut self, name: IntoString) {
        self.name = name.into();
    }

    #[inline]
    pub fn into_name(self) -> String {
        self.name
    }
}

impl<IntoString: Into<String>> From<IntoString> for Identifier {
    fn from(identifier: IntoString) -> Self {
        Self {
            name: identifier.into(),
            token: None,
        }
    }
}
