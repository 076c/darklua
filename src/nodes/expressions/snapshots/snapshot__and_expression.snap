---
source: src/nodes/expressions/binary.rs
expression: expression.to_lua_string()
---
true and false
